src/lib/y2storage/autoinst_proposal.rb

```diff
@@ -0,0 +1,184 @@
+    # @raise [NoDiskSpaceError] if there is no enough space to perform the installation
+    def propose
+      raise UnexpectedCallError if proposed?
+
+      drives = Proposal::AutoinstDrivesMap.new(initial_devicegraph, partitioning)
+
+      space_maker = Proposal::AutoinstSpaceMaker.new(disk_analyzer)
+      devicegraph = space_maker.provide_space(initial_devicegraph, drives)
+
+      @proposed_devicegraph = propose_devicegraph(devicegraph, drives)
+      @proposed = true
+
+      nil
+    end
+
+  protected
+
+    # Proposes a devicegraph based on given drives map
+    #
+    # This method falls back to #proposed_guided_devicegraph when the device map
+    # does not contain any partition.
+    #
+    # @param devicegraph [Devicegraph]                 Starting point
+    # @param drives      [Proposal::AutoinstDrivesMap] Devices map from an AutoYaST profile
+    # @return [Devicegraph] Devicegraph containing the planned devices
+    def propose_devicegraph(devicegraph, drives)
+      planned_devices = plan_devices(devicegraph, drives)
+      create_devices(devicegraph, planned_devices, drives.disk_names)
+    rescue Error
+      log.info "No partitions were specified. Falling back to guided setup planning."
+      propose_guided_devicegraph(devicegraph, drives)
+    end
+
+    # Finds a suitable devicegraph using the guided proposal approach
+    #
+    # If the :desired target fails, it will retry with the :min target.
+    #
+    # @param devicegraph [Devicegraph]       Starting point
+    # @param drives      [AutoinstDrivesMap] Devices map from an AutoYaST profile
+    # @return [Devicegraph] Proposed devicegraph using the guidede proposal approach
+    #
+    # @raise [Error] No suitable devicegraph was found
+    # @see proposed_guided_devicegraph
+    def propose_guided_devicegraph(devicegraph, drives)
+      guided_devicegraph_for_target(devicegraph, drives, :desired)
+    rescue Error
+      guided_devicegraph_for_target(devicegraph, drives, :min)
+    end
+
+    # Calculates list of planned devices
+    #
+    # If the list does not contain any partition, then it follows the same
+    # approach than the guided proposal
+    #
+    # @param devicegraph [Devicegraph]                 Starting point
+    # @param drives      [Proposal::AutoinstDrivesMap] Devices map from an AutoYaST profile
+    # @return [Array<Planned::Device>] Devices to add
+    def plan_devices(devicegraph, drives)
+      planner = Proposal::AutoinstDevicesPlanner.new(devicegraph)
+      planner.planned_devices(drives)
+    end
+
+    # Creates a devicegraph using the same approach than guided partitioning
+    #
+    # @param devicegraph [Devicegraph]       Starting point
+    # @param drives      [AutoinstDrivesMap] Devices map from an AutoYaST profile
+    # @param target      [Symbol] :desired means the sizes of the partitions should
+    #   be the ideal ones, :min for generating the smallest functional partitions
```

Comment:
it in fact return `Devicegraph` is it intention? create_devices returns it, but this one not. Confusing for me, what is intention.